src/main/java/uk/gov/justice/digital/client/s3/S3DataProvider.java

```diff
@@ -35,28 +41,94 @@ public S3DataProvider(JobArguments arguments) {
+        StructType inferredSchema = sparkSession.read().parquet(fileGlobPath).schema();
         StructType schema = withMetadataFields(sourceReference.getSchema());
-        logger.info("Schema for {}.{}: \n{}", sourceName, tableName, schema.treeString());
+        logger.info("Provided schema for {}.{}: \n{}", sourceName, tableName, schema.treeString());
+        logger.info("Inferred streaming source schema for {}.{}: \n{}", sourceName, tableName, inferredSchema.treeString());
+        logger.info("Initialising S3 data source for {}.{} with file glob path {}", sourceName, tableName, fileGlobPath);
+        if(schemasMatch(inferredSchema, schema)) {
+            return sparkSession
+                    .readStream()
+                    .schema(schema)
+                    .parquet(fileGlobPath);
+        } else {
+            String msg = "Provided and inferred schemas do not match";
+            logger.warn(msg);
+            throw new SchemaMismatchException(msg);
+        }
+    }
+
+    public Dataset<Row> getStreamingSourceDataWithSchemaInference(SparkSession sparkSession, String sourceName, String tableName) {
+        String tablePath = tablePath(arguments.getRawS3Path(), sourceName, tableName);
+
+        String fileGlobPath = ensureEndsWithSlash(tablePath) + arguments.getCdcFileGlobPattern();
+        StructType schema = sparkSession.read().parquet(fileGlobPath).schema();
+        logger.info("Inferred schema for {}.{}: \n{}", sourceName, tableName, schema.treeString());
         logger.info("Initialising S3 data source for {}.{} with file glob path {}", sourceName, tableName, fileGlobPath);
         return sparkSession
                 .readStream()
                 .schema(schema)
                 .parquet(fileGlobPath);
     }
 
-    public Dataset<Row> getSourceDataBatch(SparkSession sparkSession, SourceReference sourceReference, List<String> filePaths) {
+    public Dataset<Row> getBatchSourceData(SparkSession sparkSession, SourceReference sourceReference, List<String> filePaths) throws SchemaMismatchException {
+        val scalaFilePaths = JavaConverters.asScalaIteratorConverter(filePaths.iterator()).asScala().toSeq();
+        StructType inferredSchema = sparkSession.read().parquet(scalaFilePaths).schema();
         StructType schema = withMetadataFields(sourceReference.getSchema());
+        logger.info("Provided schema for {}.{}: \n{}", sourceReference.getSource(), sourceReference.getTable(), schema.treeString());
+        logger.info("Inferred batch source schema for {}.{}: \n{}", sourceReference.getSource(), sourceReference.getTable(), inferredSchema.treeString());
+        if(schemasMatch(inferredSchema, schema)) {
+            return sparkSession
+                    .read()
+                    .schema(schema)
+                    .parquet(scalaFilePaths);
+        } else {
+            String msg = "Provided and inferred schemas do not match";
+            logger.warn(msg);
```

Comment:
Would it be helpful for debugging/analysis to include the schemas in the log message?